README.md
# KC-Live-OS
Key Ceremony Live Operating System


Ceremony | ICANN Keytools Version         | OS/DVD Version
-------- | ------------------------------ | -----------------	
KSK 30	 |  | 
KSK 29	 | icann-keytools-20170403.tar.gz | KC-20170403.iso
KSK 28	 | icann-keytools-20161012.tar.gz | KC-20161014.iso
KSK 27	 | icann-keytools-20161012.tar.gz | KC-20161014.iso
KSK 26	 | icann-keytools-20160419.tar.gz | KC-20160503.iso
KSK 25	 | icann-keytools-20160419.tar.gz | KC-20160503.iso
KSK 24	 | icann-keytools-20110921.tar.gz | KC-20110927-r600.iso
KSK 23   | icann-keytools-20110921.tar.gz | KC-20110927-r600.iso
KSK 22	 | icann-keytools-20110921.tar.gz | KC-20110927-r600.iso
KSK 21	 | icann-keytools-20110921.tar.gz | KC-20110927-r600.iso
KSK 20	 | icann-keytools-20110921.tar.gz | KC-20110927-r600.iso
KSK 19	 | icann-keytools-20110921.tar.gz | KC-20110927-r600.iso
KSK 18	 | icann-keytools-20110921.tar.gz | KC-20110927-r600.iso
KSK 17	 | icann-keytools-20110921.tar.gz | KC-20110927-r600.iso
KSK 16	 | icann-keytools-20110921.tar.gz | KC-20110927-r600.iso
KSK 15	 | icann-keytools-20110921.tar.gz | KC-20110927-r600.iso
KSK 14	 | icann-keytools-20110921.tar.gz | KC-20110927-r600.iso
KSK 13	 | icann-keytools-20110921.tar.gz | KC-20110927-r600.iso
KSK 12	 | icann-keytools-20110921.tar.gz | KC-20110927-r600.iso
KSK 11	 | icann-keytools-20110921.tar.gz | KC-20110927-r600.iso
KSK 10	 | icann-keytools-20110921.tar.gz | KC-20110927-r600.iso
KSK 09	 | icann-keytools-20110921.tar.gz | KC-20110927-r600.iso
KSK 08	 | icann-keytools-20110921.tar.gz | KC-20110927-r600.iso
KSK 07	 | icann-keytools-20110921.tar.gz | KC-20110927-r600.iso
KSK 06	 | icann-keytools-20110921.tar.gz | KC-20110927-r600.iso
KSK 05	 | icann-keytools-20101029.tar.gz | KC-20101029-r575.iso
KSK 04	 | icann-keytools-20101029.tar.gz | KC-20101029-r575.iso
KSK 03	 | icann-keytools-20101029.tar.gz | KC-20101029-r575.iso
KSK 02	 | icann-keytools-20100612.tar.gz | KC-20100612-r482.iso
KSK 01	 | icann-keytools-20100612.tar.gz | KC-20100612-r482.iso


ChangeLog
2017-04-03	KSK Rollover
	
	Version 20170403

	a. Added source code and checksum for ICANN Keytools 20170403.
	b. Updated KSR Signer.
	c. Updated printlog.
	d. Added hsmfd-hash.
	e. Updated /etc/localtime to UTC
	f. Updated O/S DVD Version
	g. Updated /etc/SERIAL file.


2016-10-14	RSSAC003 Recommendation

	Version 20161014

	a. Added source code and checksum for ICANN Keytools 20161012.
	b. Added tool sha2wordlist to calculate a SHA-256 checksum and corresponding PGP 
	   words for a file. This utility has a dependency on libbsd.so, a library providing 
	   useful functions commonly found on BSD systems.
	c. Updated KSK Generator.
	d. Updated KSR Signer.
	e. Updated O/S DVD Version.
	f. Updated /etc/SERIAL file.


2016-05-03	ZSK Length Change

	Version 20160503

	a. Added source code and checksum for ICANN Keytools 20160419.
	b. Updated KSK Generator.
	c. Updated O/S DVD Version.
	d. Updated /etc/SERIAL file.


2011-09-27	180 Days Warning 

	Version 20110927-r600

	a. Added source code and checksum for ICANN Keytools 20110921.
	b. Updated binaries.


2010-10-29	x509 Schema

	Version 20101029-r575

	a. Added source code and checksum for ICANN Keytools 20101029.
	b. Updated binaries.
	c. Updated utilities.


2010-06-12	Original

	Version 20100612-r482

	First version.
	
	a. Added source code and checksum for ICANN Keytools 20100612.
	b. Added binaries.
	c. Added utilities.




Centos/create-kc-iso.sh
#!/bin/sh -x
#
# Creating a new KC DVD ISO based on the previous KC DVD ISO
# Adding SW components, recompress it into a SquashFS and create the new ISO. 
#
# To-do:
# Define the path for the working directory and also for the ISO. Maybe $1 and $2 
# Write the warning and consideration to run the scrip - need to be run in a Linux machine, recommendation a minimal virtual machine
# Verify the signature
# Sign this scrip
# Any way to check the scrip version?
# Verify the final hash of the ISO
# Add a clean up
# wget *.iso
# wget the custom KM-SW?, and all the other files?
# check if automount is disable
# make for any unix/linux distribution
# Verify the mksquashfs version 
# Instead of moving, copy and deleting the squashfs.img just unsquashfs from the original directory, them replace the old for the new.
#
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. 

# Be sure to verify that you will have sufficient space before running the script.
# This may require, for example, over 12.0 GiB of free space for Fedora 25 Workstation Live

# *CHANGE* #
##KCDVD_VERSION="KC-20161014.iso" # Current KC DVD ISO version
##KCDVD_HASH="991f7be8cfbc3b4bdb6f5e5f84092486755a08a3c36712e37a26ccd808631692" # SHA-256 of KC-20161014.iso
M_ISO="KC-20161014"		# Mount folder for the iso
# *CHANGE* #

ROOT_UID=0	# Only users with $UID 0 have root privileges
M_WD="KC-$(date +%Y%m%d)"	# Working directory to create the new ISO version = today
M_SMNT=squashmnt        	# Mount folder for mounting squash file system 
M_SFS=squashfs          	# Mount folder for squash file system
M_RFS=rootfs            	# Mount folder for root ext3fs file system 

# Update this function accordingly to the O/S changes
update_ext3fs()
{
# Entering to chroot environment
# Changing time zone just for this time!
chroot $M_RFS ln -fs /usr/share/zoneinfo/UTC /etc/localtime > /dev/null 2>&1
#ln -fs /usr/share/zoneinfo/UTC $M_RFS/etc/localtime 
 
# Updating /opt/icann/bin ksrsigner
##chown root:root ksrsigner
##chmod 555 ksrsigner
##if [[ ! sha256sum -c ksrsigner.sha256 ]]
##then
##	exit 1
##fi
##rsync -ar --inplace --progress ksrsigner $M_RFS/opt/icann/bin/
##if [[ ($(sha256sum $M_RFS/opt/icann/bin/ksrsigner)) != awk '{print $1}' ksrsigner.sha256 ]]
##then
##	exit 1
##fi

# *CHANGE* #
# for all binaries http://www.iitk.ac.in/LDP/LDP/abs/html/abs-guide.html#EX22A
# list is a file name and chmod permission
for bin in "ksrsigner 555" "printlog 755" "hsmfd-hash 755"
do
	set -- $bin # parses variable "bin"
	chown root:root $1
	chmod $2 $1
	#if ! sha256sum -c $1.sha256
	if [[ ! "$(sha256sum -c $1.sha256)" ]]
	then
		exit 1
	fi
	rsync -ar --inplace --progress $1 $M_RFS/opt/icann/bin/
	sha1=($(sha256sum $M_RFS/opt/icann/bin/$1))
	sha2=$(awk '{print $1}' $1.sha256)
	if [[ "$sha1" != "$sha2" ]]
	#if [[ "($(sha256sum $M_RFS/opt/icann/bin/$1))" != "$(awk '{print $1}' $1.sha256)" ]]
	then
		exit 1
	fi
done
}

# Confirmation source: http://stackoverflow.com/questions/1885525/how-do-i-prompt-a-user-for-confirmation-in-bash-script
echo "Warning this can be dangerous. If something is going wrong can change from your host system rather than from the Live CD image. You need to be root and execute under your own responsibility"
read -n 1 -r -p "Are you sure to continue [y/N]? " continue
echo    # Move to a new line
if [[ ! $continue =~ ^[Yy]$ ]]
then
	exit 1
fi

# Run as root source: http://www.iitk.ac.in/LDP/LDP/abs/html/abs-guide.html#EX2
if [ "$UID" -ne "$ROOT_UID" ]
then
  echo "Must be root to run this script."
  exit 1
fi

# Checking squashfs-tools source: http://stackoverflow.com/questions/592620/check-if-a-program-exists-from-a-bash-script
# ADD check VERSION
command -v mksquashfs >/dev/null 2>&1 || { echo >&2 "Please install squashfs-tools"; exit 1; }

# Checking rsync
command -v rsync >/dev/null 2>&1 || { echo >&2 "Please install rsync"; exit 1; }

# Checking xorriso
# Not in CentOS 5.5 command -v rsync >/dev/null 2>&1 || { echo >&2 "Please install xorriso"; exit 1; }

# Checking the ISO HASH
##echo "Calculating the ISO SHA-256 HASH of the $KCDVD_VERSION"
##iso_hash=($(sha256sum $KCDVD_VERSION))
##echo "SHA-256 HASH: $iso_hash"
##echo "KC-DVD  HASH: $KCDVD_HASH"
##if [ "$iso_hash" != "$KCDVD_HASH" ]
##then
##	echo "ERROR: SHA-256 hashes do not match, try to download again the $KCDVD_VERSION"
##	exit 1
##else
##	echo "SHA-256 HASH of the $KCDVD_VERSION match"
##fi
##
if [[ ! "$(sha256sum -c $M_ISO.iso.sha256)" ]]
then
	exit 1
fi

# Using OSIRROX a more efficient way to copy the iso content
##xorriso -osirrox on -indev $KCDVD_VERSION -extract / $M_WD
# Mounting the ISO
mkdir $M_ISO
mount -o loop $M_ISO.iso $M_ISO

# Creating the work directory
mkdir $M_WD

# Coping the entire ISO
rsync -ar --inplace --progress $M_ISO/ $M_WD
##cp -rdav $M_ISO/* $M_WD/ #rsync is better with progress information 

# umount the ISO it not longer needed
umount $M_ISO
rmdir $M_ISO

# Changing isolinux.cfg O/S DVD Version
sed -i s/$M_ISO/$M_WD/g $M_WD/isolinux/isolinux.cfg

# Moving the squashfs.img to current directory
mv $M_WD/LiveOS/squashfs.img . 

# Using unsquashfs a more efficient way to copy the file system 
# Mounting the squashfs.img
#mkdir $M_SMNT
#mount -o loop -t squashfs squashfs.img $M_SMNT
#
# Copy squashfs file system
#mkdir $M_SFS
#rsync -ar --inplace --progress $M_SMNT/ $M_SFS
#
# Umount the squashfs.img it not longer needed
#umount $M_SMNT
#rmdir $M_SMNT

unsquashfs -dest $M_SFS squashfs.img 

# Removing the squashfs.img
rm -f squashfs.img

# Just for this time, changing ext3fs.img owner to root
chown root:root $M_SFS/LiveOS/ext3fs.img

# Mounting the root root file system
mkdir $M_RFS
mount -o loop $M_SFS/LiveOS/ext3fs.img $M_RFS

# Edit section 
update_ext3fs

# Adding the icann-keytools
chown root:root icann-keytools-*
chmod 644 icann-keytools-*
if [[ ! "$(sha256sum -c icann-keytools-*.sha256)" ]]
then
	exit 1
fi
rsync -ar --inplace --progress icann-keytools-* $M_RFS/opt/icann/dist/

#if [[ "($(sha256sum $(ls -tr $M_RFS/opt/icann/dist/icann-keytools-*.tar.gz | tail -n 1)))" != "$(awk '{print $1}' icann-keytools-*.sha256)" ]]
sha1=($(sha256sum $(ls -tr $M_RFS/opt/icann/dist/icann-keytools-*.tar.gz | tail -n 1)))
sha2=$(awk '{print $1}' icann-keytools-*.sha256)
if [[ "$sha1" != "$sha2" ]]
then
	exit 1
fi

# SERIAL
sed -i s/$M_ISO/$M_WD/g $M_RFS/etc/SERIAL

# Umount the root file system
umount $M_RFS
rmdir $M_RFS

# Creating the new squashfs, may want to use XZ compression
mksquashfs $M_SFS/ squashfs.img -noappend

# Carefully removing the squash file system
rm -rf $M_SFS

# Moving the squashfs.img to working directory
mv squashfs.img $M_WD/LiveOS/ 

# Setting permissions for squashfs.img
chmod 555 $M_WD/LiveOS/squashfs.img

# Creating the iso
mkisofs -J -l -r -cache-inodes --hide-rr-moved -hide-joliet-trans-tbl \
	-V $M_WD -o $M_WD.iso -b isolinux/isolinux.bin -c isolinux/boot.cat \
	-no-emul-boot -boot-info-table -boot-load-size 4 $M_WD

# Carefully removing working directory
rm -rf $M_WD

# Creating the new iso HASH
sha256sum $M_WD.iso > $M_WD.iso.sha256

# END


Fedora/create-kc-iso-Fedora.sh
#!/bin/bash -x
#
# Creating the KC DVD ISO based on a Live CD.
# Changes, remove, recompress it into a SquashFS image, add, hash, etc 
#
# To-do:
# Define the path for the working directory and also for the ISO. Maybe $1 and $2 
# Write the warning and consideration to run the scrip - need to be run in a Linux machine, recommendation a minimal virtual machine
# Verify the signature with Fedora PGP
# Sign this scrip
# Any way to check the scrip version?
# Verify the final hash of the ISO
# Add a clean up
# wget fedora.iso
# wget the custom KM-SW?, and all the other files?
# check if automount is disable
# make for any unix/linux distribution
# Verify the mksqushfd version 
# Instead of moving, copy and deleting the squashfs.img just unsquashfs from the original directory, them replace the old for the new.
# Use only one source of time and use a tool to change it
#
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. 

# Be sure to verify that you will have sufficient space before running the script.
# This may require, for example, over 12.0 GiB of free space for Fedora 25 Workstation Live


ROOT_UID=0	# Only users with $UID 0 have root privileges
FEDORA_VERSION="Fedora-Workstation-Live-x86_64-25-1.3.iso"	# Current fedora live version
FEDORA_HASH="818017f42a2741cfaf20e94aecf6a63d1b995abfdaff5917df7218d0d89976a7  -" # the " -" is necessary for the comparative unless if "sed" is used SHA-256 of (Fedora-Workstation-Live-x86_64-25-1.3.iso) 
KS_DVD=0	# SHA256 ()
M_ISO=live-iso	# Mount folder for the iso
M_WD=KC-20170401	# Working directory to create the ISO
M_SMNT=squashmnt	# Mount folder for mounting squash file system
M_SFS=squashfs		# Mount folder for squash file system
M_RFS=rootfs		# Mount folder for root file system
B_TIME=1491004800       # Time for squashfs https://www.epochconverter.com/ Human time (GMT): Sat, 01 Apr 2017 00:00:00 GMT
M_TIME=2017040100000000 # Time YYYYMMDDhhmmsscc to control the timestamps of the filesystem superblocks and other global components of the ISO file system
F_TIME=201704010000.00  # Ext3 Filesystem time

# Confirmation source: http://stackoverflow.com/questions/1885525/how-do-i-prompt-a-user-for-confirmation-in-bash-script
echo "Warning this can be dangerous. It will use chroot command to remove packages, changes configurations, etc. So, if something is going wrong can change from your host system rather than from the Live CD image. You need to be root and execute under your own responsibility"
read -p "Are you sure to continue [y/N]? " -n 1 -r
echo    # Move to a new line
if [[ ! $REPLY =~ ^[Yy]$ ]]
then
    [[ "$0" = "$BASH_SOURCE" ]] && exit 1 || return 1 # handle exits from shell or function but don't exit interactive shell
fi

# Run as root source: http://www.iitk.ac.in/LDP/LDP/abs/html/abs-guide.html#EX2
if [ "$UID" -ne "$ROOT_UID" ]
then
  echo "Must be root to run this script."
  exit 1
fi

# Checking squashfs-tools source: http://stackoverflow.com/questions/592620/check-if-a-program-exists-from-a-bash-script
# ADD check VERSION
command -v mksquashfs >/dev/null 2>&1 || { echo >&2 "Please install (with XZ support) the last squashfs-tools direct form GitHub https://github.com/squashfs-tools/squashfs-tools.git"; exit 1; }

#unsquashfs version 4.3 (2014/05/12)
#mksquashfs version 4.3-git (2014/09/12)

# Checking rsync
command -v rsync >/dev/null 2>&1 || { echo >&2 "Please install rsync"; exit 1; }

# Checking xorriso
command -v xorriso >/dev/null 2>&1 || { echo >&2 "Please install xorriso"; exit 1; }

#xorriso version   :  1.4.6
#Version timestamp :  2016.09.16.133001
#Build timestamp   :  -none-given-
#libisofs   in use :  1.4.6  (min. 1.4.6)
#libburn    in use :  1.4.6  (min. 1.4.6)
#libburn OS adapter:  internal GNU/Linux SG_IO adapter sg-linux
#libisoburn in use :  1.4.6  (min. 1.4.6)

# Checking the ISO HASH
echo "Calculating the ISO SHA-256 HASH of the $FEDORA_VERSION"
iso_hash=$(sha256sum < "$FEDORA_VERSION") # | sed 's/\(.*\) .*/\1/') # in the same directory as the script
echo "SHA-256 HASH: $iso_hash"
echo "Fedora  HASH: $FEDORA_HASH"
if [ "$iso_hash" != "$FEDORA_HASH" ]
then
  echo "ERROR: SHA-256 hashes mismatched, try to download again the $FEDORA_VERSION"
  exit 1
else
  echo "SHA-256 HASH of the $FEDORA_VERSION is OK"
fi

# Using xorriso with osirrox a more efficient way to copy the iso content
## Mounting the Fedora Live ISO
#mkdir $M_ISO
#mount -o loop $FEDORA_VERSION $M_ISO
#
# Creating the work directory
mkdir $M_WD
#
## Coping the entire ISO
#rsync -ar --inplace --progress $M_ISO/ $M_WD
###cp -rdav $M_ISO/* $M_WD/ #rsync is better with progress information 
#
## umount the ISO it not longer needed
#umount $M_ISO
#rmdir $M_ISO

xorriso -osirrox on -indev $FEDORA_VERSION -extract / $M_WD

# Changing isolinux.cfg
# Reducing the boot menu time
sed -i 2s/6/1/ $M_WD/isolinux/isolinux.cfg

# Verbose booting output
sed -i 64s/quiet/#quiet/ $M_WD/isolinux/isolinux.cfg

# Moving the squashfs.img to current directory
mv $M_WD/LiveOS/squashfs.img . 

# Using unsquashfs a more efficient way to copy the file system 
# Mounting the squashfs.img
#mkdir $M_SMNT
#mount -o loop -t squashfs squashfs.img $M_SMNT
#
# Copy squashfs file system
mkdir $M_SFS
#rsync -ar --inplace --progress $M_SMNT/ $M_SFS
#
# Umount the squashfs.img it not longer needed
#umount $M_SMNT
#rmdir $M_SMNT

unsquashfs -f -d $M_SFS squashfs.img 

# Removing the squashfs.img
rm -f squashfs.img

# Mounting the root root file system
mkdir $M_RFS
mount $M_SFS/LiveOS/rootfs.img $M_RFS

# Edit section 
# Disabling selinux
sed -i 7s/enforcing/disabled/ $M_RFS/etc/selinux/config

# Setting network

# ADD MORE

# Entering to chroot environment
# chroot $M_RFS  

# Disabling firewall
chroot $M_RFS systemctl disable firewalld > /dev/null 2>&1
chroot $M_RFS systemctl mask firewalld > /dev/null 2>&1
#rm -f $M_RFS/etc/systemd/system/dbus-org.fedoraproject.FirewallD1.service
#rm -f $M_RFS/etc/systemd/system/multi-user.target.wants/firewalld.service
#ln -s /dev/null $M_RFS/etc/systemd/system/firewalld.service

# Remove unnecessary packages 
#chroot $M_RFS dnf -y remove firefox libreoffice-core cheese evolution gnome-contacts gnome-maps rhythmbox shotwell gnome-weather > /dev/null 2>&1
# dnf add a timestamp inside of the RPM database, ufff

# Changing time zone
chroot $M_RFS ln -fs /usr/share/zoneinfo/UTC /etc/localtime > /dev/null 2>&1
#ln -fs /usr/share/zoneinfo/UTC $M_RFS/etc/localtime 

# Seting filesystem timestamp
for file in $(find "$M_RFS" -mtime 0)
do
	touch -a -m -h -t $F_TIME $file
done 

# Umount the root file system
umount $M_RFS
rmdir $M_RFS

# Creating the new squashfs, may want to use XZ compression
mksquashfs $M_SFS/ squashfs.img -noappend -comp xz -mkfs-fixed-time $B_TIME -content-fixed-time $B_TIME

# Carefully removing the squash file system
rm -rf $M_SFS

# Moving the squashfs.img to working directory
mv squashfs.img $M_WD/LiveOS/ 

# Setting permissions for squashfs.img
chmod 644 $M_WD/LiveOS/squashfs.img

# Change the timestamp for the modified files 
for files in $(find "$M_WD" -mtime 0)
do
	touch -a -m -h -t $F_TIME $files
done 

## Creating the iso
#xorriso -as mkisofs -joliet -full-iso9660-filenames -rational-rock \
#	-hide-rr-moved -volid ${M_WD/-/_} -output $M_WD.iso \
#	-eltorito-boot isolinux/isolinux.bin -eltorito-catalog isolinux/boot.cat \
#	-no-emul-boot -boot-info-table -boot-load-size 4 \
#	modification-date=$M_TIME \
#	$M_WD
#
## Carefully removing working directory
#rm -rf $M_WD

# Checking the new iso HASH

# END 
